kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/ImplicitCastExpression.java

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal.expressions;
+
+import java.util.*;
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.expressions.Expression;
+import io.delta.kernel.types.DataType;
+
+import io.delta.kernel.defaults.client.DefaultExpressionHandler;
+
+/**
+ * An implicit cast expression to convert the input type to another given type. Here is the valid
+ * list of casts
+ * <p>
+ *  <ul>
+ *    <li>{@code byte} to {@code short, int, long, float, double}</li>
+ *    <li>{@code short} to {@code int, long, float, double}</li>
+ *    <li>{@code int} to {@code long, float, double}</li>
+ *    <li>{@code long} to {@code float, double}</li>
+ *    <li>{@code float} to {@code double}</li>
+ *  </ul>
+ *
+ * <p>
+ * The above list is not exhaustive. Based on the need, we can add more casts.
+ * <p>
+ * In {@link DefaultExpressionHandler} this is used when the operands of an expression are not of
+ * the same type, but the evaluator expects same type inputs. There could be more use cases, but
+ * for now this is the only use case.
+ */
+final class ImplicitCastExpression implements Expression {
```

Comment:
Want to create a github issue to follow up on this? Seems pretty weird that a comparison between decimal types with different scales (which just determines how it's written in parquet and not the actual value) would fail

Reply:
https://github.com/delta-io/delta/issues/2044 (also mentioned in the PR description)